Automattic-Tracks-iOS/CrashLogging.swift

```diff
@@ -0,0 +1,156 @@
+import Foundation
+import Sentry
+
+/// A class that provides support for logging crashes. Not compatible with Objective-C.
+public class CrashLogging {
+
+    /// A singleton is maintained, but the host application needn't be aware of its existence.
+    internal static let sharedInstance = CrashLogging()
+    fileprivate var dataProvider: CrashLoggingDataProvider?
+    
+    /**
+     Initializes the crash logging system.
+
+     - Parameters:
+     - dataProvider: An object that will provide any required data to the crash logging system.
+
+     - SeeAlso: CrashLoggingDataProvider
+     */
+    public static func start(withDataProvider dataProvider: CrashLoggingDataProvider) {
+
+        // Store the data provider for future use
+        sharedInstance.dataProvider = dataProvider
+
+        // Create a Sentry client and start crash handler
+        do {
+            Client.shared = try Client(dsn: dataProvider.sentryDSN)
+            try Client.shared?.startCrashHandler()
+
+            // Store lots of breadcrumbs to trace errors
+            Client.shared?.breadcrumbs.maxBreadcrumbs = 500
+
+            // Automatically track screen transitions
+            Client.shared?.enableAutomaticBreadcrumbTracking()
+
+            // Automatically track low-memory events
+            Client.shared?.trackMemoryPressureAsEvent()
+
+            // Override event serialization to append the logs, if needed
+            Client.shared?.beforeSerializeEvent = sharedInstance.beforeSerializeEvent
+            Client.shared?.shouldSendEvent = sharedInstance.shouldSendEvent
+
+            // Add additional data
+            Client.shared?.releaseName = dataProvider.releaseName
+            Client.shared?.environment = dataProvider.buildType
+            updateCurrentUser()
+
+        } catch let error {
+            logError(error)
+        }
+    }
+
+    /// A Sentry hook used to attach any additional data to the event.
+    private func beforeSerializeEvent(_ event: Event) {
+        event.tags?["locale"] = NSLocale.current.languageCode
+    }
+
+    /// A Sentry hook that controls whether or not the event should be sent.
+    private func shouldSendEvent(_ event: Event?) -> Bool {
+
+        #if DEBUG
+        let result = false
+        #else
+        let result = !CrashLogging.userHasOptedOut
+        #endif
+
+        shouldSendEventCallback?(result)
+
+        return result
+    }
+
+    /// The current state of the user's choice to opt out of data collection. Provided by the data source.
+    public static var userHasOptedOut: Bool {
```

Comment:
❓Thanks for bringing this one up! So, by making this `private`, you're thinking it's better if the application developer reads this from their own `CrashLoggingDataProvider` implementation (or whatever that's calling underneath) rather than off this global class?

I can definitely be convinced of this – I left it `public` as a handy global lookup, but I think I like the simplicity of closing it off, especially since, as you pointed out, in a well-architected app it shouldn't be needed.

Reply:
>  you're thinking it's better if the application developer reads this from their own CrashLoggingDataProvider implementation (or whatever that's calling underneath) rather than off this global class?

Somewhat. I believe it's **not** the Crash Logger's responsibility to track this setting. It should be somewhere in the specific app's `UserPreferences` class or whatever they call it. 

`CrashLogging` should probably even call this property as `loggingEnabled: Bool` and doesn't care that it's `false` because the user opted out.

Reply:
Privatized in 65186d190954e93f2254aef8a174f169b3dd7d16!